config/config/src/main/java/io/helidon/config/EnumMapperProvider.java

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ */
+package io.helidon.config;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import io.helidon.config.spi.ConfigMapperProvider;
+
+/**
+ * Built-in mapper for {@code enum}s.
+ *
+ * <p>
+ *     This mapper attempts to match strings in the config source to enum values as follows:
+ *     <ul>
+ *         <li>The mapper treats hyphens ('-') in config strings as underscores when comparing to enum value names.</li>
+ *         <li>If the matcher finds a <em>case-sensitive</em> match with an enum value name, then that enum value matches.</li>
+ *         <li>If the matcher finds exactly one <em>case-insensitive</em> match, that enum value matches.</li>
+ *         <li>If the matcher finds no matches or multiple matches, throw a
+ *         {@link io.helidon.config.ConfigEnumMappingException} with a message explaining the problem.</li>
+ *     </ul>
+ *     These conversions are intended to maximize ease-of-use for authors of config sources so the values need not be
+ *     upper-cased nor punctuated with underscores rather than the more conventional (in config at least) hyphen.
+ * </p>
+ * <p>
+ *     The only hardship this imposes is if a confusingly-designed enum has values which differ only in case <em>and</em> the
+ *     string in the config source does not exactly match one of the enum value names. In such cases
+ *     the mapper will be unable to choose which enum value matches an ambiguous string. A developer faced with this
+ *     problem can simply provide her own explicit config mapping for that enum, for instance as a function parameter to
+ *     {@code Config#as}.
+ * </p>
+ *
+ */
+class EnumMapperProvider implements ConfigMapperProvider {
+
+    private final Map<Class<?>, Function<Config, ?>> mappers = new HashMap<>();
+
+    @Override
+    public Map<Class<?>, Function<Config, ?>> mappers() {
+        return mappers;
+    }
+
+    @Override
+    public <T> Optional<Function<Config, T>> mapper(Class<T> type) {
+        if (!type.isEnum()) {
+            return Optional.empty();
+        }
+
+        Function<Config, T> result = (Function<Config, T>) mappers.computeIfAbsent(type, this::enumMapper);
+        return Optional.of(result);
+    }
+
+    private <T> Function<Config, T> enumMapper(Class<T> type) {
+        return config -> {
+            // type.isEnum() must be true in the caller, so we know the type is Class<Enum<?>>.
+            Class<Enum<?>> enumType = (Class<Enum<?>>) type;
+            if (!config.hasValue() || !config.exists()) {
+                throw MissingValueException.create(config.key());
+            }
+            if (!config.isLeaf()) {
+                throw new ConfigMappingException(config.key(), enumType, "config node must be a leaf but is not");
+            }
+            String value = config.asString().get();
```

Comment:
why did you decide to do the conversion rules? 
I though this would be a simple use of the value (I could also understand the case insensitiveness). 
I am not much in favor of doing tricks like this to try to "handle all cases". If they want "MY_VALUE", they can specify "MY_VALUE" or "my_value"... no need to support dashes and/or underscores (you could come up with lot of rules then, which is not very helpful)

Reply:
I described my thinking in the JavaDoc for `EnumMapperProvider`:

 > These conversions are intended to maximize ease-of-use for authors of config sources so the values need not be   upper-cased nor punctuated with underscores rather than the more conventional (in config at least) hyphen.

I generally agree about not trying to be overly clever in code trying to guess what a user intends, particularly where that intent is ambiguous. Here, though, the purpose is definitely _not_ to "handle all cases" but to anticipate two specific mistakes I expect users will commonly make where their intent is quite clear:
* The hyphen is not a legal character in a Java `enum` value declaration. Given the appearance of a hyphen in a config value where an `enum` value name is expected, the user intent seems clear.
* Case insensitivity seemed obvious (and we agree on that).

I suggested these two specific rules--not "lot of rules"--to help users avoid two clear sources of errors and frustration (which seem likely) and help us avoid subsequent incoming support traffic.

Most importantly for users, I think the conversion rules improve ease-of-use without really adding confusion or ambiguity in how Helidon will behave.

Reply:
OK, let's keep it in, as hyphen is not legal in enum value declarations.